### PR TITLE
Set up environments and lint and formatter

### PR DESCRIPTION
**Pull request by ZidongLiu**
# Summary

- [x] README update, will do later with @RogerRogerUSC, will include how to install, and how to set up ruff locally with vscode
- [x] Get environment file for conda, (cuda packages are platform special, and not available on mac)
- [x] Add ruff/pyright to env
- [x] Add peft
- [x] CI using conda
- [x] Add ruff lint check to ci
- [x] Add ruff format check to ci
- [x] Ruff format locally

Try to install env using following command:

```
conda env create -f environment.yml -y
conda activate decomfl
```

Code still works
image

### CODE REVIEW

File: slurm_main.sh [hunk 1]
```diff
@@ -1,18 +0,0 @@
-#!/bin/bash -l
```
Comment (BichengYing) [inline slurm_main.sh:1]: why do you delete this file?

File: environment.yml [hunk 2]
```diff
@@ -0,0 +1,22 @@
+name: decomFL
+channels:
+  - pytorch
+  - conda-forge
+  - defaults
+dependencies:
+  - python=3.10.14
```
Comment (BichengYing) [inline environment.yml:7]: there is no ~approx usage?

File: .github/workflows/ci.yaml [hunk 3]
```diff
@@ -1,33 +1,39 @@
...
 on:
   push:
-    branches: [ "main" ]
+    branches: ["main"]
   pull_request:
-    branches: [ "main" ]
-
-permissions:
-  contents: read
+    branches: ["main"]
```
Comment (BichengYing) [inline .github/workflows/ci.yaml:7]: why not trigger the ci for all branches?

Did you turn off the permission of direct push to main branch?

**Review by BichengYing — APPROVED** [review]
LGTM! This is a  huge step!

Not in this PR, I think the only main missing piece is Mypy (or equivalent static type check)

File: .github/workflows/ci.yaml (hunk 3 above)
Comment (ZidongLiu) [reply .github/workflows/ci.yaml:7]: image
direct push is disabled

File: environment.yml [hunk 4]
```diff
@@ -0,0 +1,24 @@
+name: decomFL
```
[the remark this answers is not in the document]
Comment (ZidongLiu) [reply environment.yml:1]: not sure, you can decide, will adjust the environment name accordingly

File: slurm_main.sh (hunk 1 above)
Comment (ZidongLiu) [reply slurm_main.sh:1]: feel rit computing resource is not in need at this moment. want to clean up our root files. Will get them back in a later PR after code refactoring

File: environment.yml (hunk 2 above)
Comment (ZidongLiu) [reply environment.yml:7]: we can pin 3.10, but I remember @RogerRogerUSC mentioned in wechat we are using 3.10.14

File: environment.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (BichengYing) [reply environment.yml:1]: use all lower case then (easier to type)

File: environment.yml (hunk 2 above)
Comment (BichengYing) [reply environment.yml:7]: either way are fine

File: environment.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (ZidongLiu) [reply environment.yml]: use `decomfl`

File: environment.yml (hunk 2 above)
Comment (ZidongLiu) [reply environment.yml]: pin 3.10 instead of 3.10.14

File: .github/workflows/ci.yaml (hunk 3 above)
Comment (ZidongLiu) [reply .github/workflows/ci.yaml:7]: will turn ci on only when targeting `main` for now

**Comment by ZidongLiu** [comment]
Still works
image

File: README.md [hunk 5]
```diff
@@ -18,24 +17,100 @@ Table 1: Test accuracy comparsion of First-order SGD, MeZO and DeComFL on OPT-12
...
+
+- **Run DeComFL:** Follow FL routine, split data into chunks and train on different clients.
+  Usage example: `python decomfl_main.py --dataset=sst2 --iterations=10000 --train-batch-size=8 --test-batch-size=200 --eval-iterations=50 --num-clients=3 --num-sample-clients=2 --local-update-steps=1 --num-pert=10 --lr=1e-6 --mu=1e-3 --grad-estimate-method=rge-forward`
+
+## Code Style
 
-- **Run DeComFL:** Follow FL routine, split data into chunks and train on different clients. 
-Usage example: `python decomfl_main.py --dataset=sst2 --iterations=10000 --train-batch-size=8 --test-batch-size=200 --eval-iterations=50 --num-clients=3 --num-sample-clients=2 --local-update-steps=1 --num-pert=10 --lr=1e-6 --mu=1e-3 --grad-estimate-method=rge-forward`
+We use `Ruff==0.6` for both lint and format which should be installed in the created conda virtual env together with other non-dev dependencies.
+
```
Comment (BichengYing) [inline README.md:50]: extra new line here

File: README.md [hunk 6]
```diff
@@ -18,24 +17,100 @@ Table 1: Test accuracy comparsion of First-order SGD, MeZO and DeComFL on OPT-12
...
+We will use `pyright` for static type checking, which is on the way.
+
+If using VScode, you can rely on VScode ruff plugin. See more in next section for [VScode](#vscode).
```
Comment (BichengYing) [inline README.md:53]: this sentence is unnecessary

File: README.md [hunk 7]
```diff
@@ -18,24 +17,100 @@ Table 1: Test accuracy comparsion of First-order SGD, MeZO and DeComFL on OPT-12
...
+		"editor.formatOnPaste": false,
+    	"editor.formatOnType": false,
+		"editor.formatOnSave": true,
+		"editor.defaultFormatter": "charliermarsh.ruff",
+	},
+	"ruff.enable": true,
+	"ruff.interpreter": ["/path/to/your/conda-venv/python"], # or ["/path/to/your/conda-venv/python.exe"] for Windows
+	"ruff.path": ["/path/to/your/conda-venv/bin/ruff"], # or ["/path/to/your/conda-venv/Scripts/ruff.exe"] for Windows
+}
+```
```
Comment (BichengYing) [inline README.md:105]: we should create a folder code for `devtool`. This part belongs to that. Notice two role here -- 1) the user of our library, and 2) the developer of this repo.  This part is for developer only so it should be under this README.md, which is for user.

Comment (BichengYing) [reply README.md:105]: see https://github.com/quantumlib/Cirq as example

Comment (BichengYing) [reply README.md:105]: under the devtool, we can put some convenient script such as `ruff format @1 && ruff check @1` etc.

File: README.md (hunk 5 above)
Comment (ZidongLiu) [reply README.md:50]: This extra line is auto formatted by ruff

File: README.md (hunk 6 above)
Comment (ZidongLiu) [reply README.md]: Simplified wording

File: README.md (hunk 7 above)
Comment (ZidongLiu) [reply README.md]: Splitted into two README.md

**Review by BichengYing — APPROVED** [review]

**Pull request merged**